content/blog/2020-05-22-rt-template/index.md

```diff
@@ -0,0 +1,365 @@
+        "chain_replace", Type::getDoubleTy(Ctx),
+        Type::getDoubleTy(Ctx), Type::getDoubleTy(Ctx)
+    );
+    // function call name and argument types have to be known
+
+    for (auto &B: F) {
+        list <Instruction*> Selected;
+        for (auto &I: B) {
+        /* Add runtime sub block selection here */
+    
+        /* End of selection algorithm */
+    	}
+    
+        Value* args[4];
+        /* Enter instructions to the new basic block */
+        for (auto &I: Selected) {
+            IRBuilder<> builder(I);
+            builder.SetInsertPoint(&B, ++builder.GetInsertPoint());
+            Value* ret =  builder.CreateCall(rtFunc, args);
+    
+            /* forward return to users */
+            for (auto& U : I->uses()) {
+                User* user = U.getUser();
+                user->setOperand(U.getOperandNo(), ret);
+            }
+        }
+        /* Done entering new instructions */
+        
+        /* Remove replaced instructions from the basic block */
+        while (!Selected.empty()) {
+            Selected.front()->eraseFromParent();
+            Selected.pop_front();
+        }
+    }
+```
+The tradeoff with this approach is, this abstraction is an arbitrary one we have picked.
+The program, nor llvm provide any support. We have to create the interfaces and replacement
+with lower level tools such as basic blocks or instructions. 
+
+## Evaluation
+I evaluated this framework by implementing passes for [MachSuite][mach]. I use a mix of unoptimizes and `-O1`
+for the kernel code. I use `clang version 11.0.0` for llvm compilation and `python3.7` for Python execution. 
+I used a timing details logging to make the tests more robust. However, for certain tests this proved
+too tedious as the tests took a long time to complete.
+
+For instructions, I replaced floating point instructions with a python simulator of
+a floating point accelerator. At this flexible granularity it was fairly straight-forward to cover 
+the entire benchmark. 
+
+For functions, I created a vvadd function to test replacement for a vvadd benchmark. I also created 
+a preprocessed results reading function for functions with 2 input arguments and an output argument of 
+the same size. This function can only cover 3 microbenchmarks I wrote and 2 MachSuite benchmarks as 
+the function signature doesn't match rest of the suite. Extending the runtime library is fairly trivial,
+but incomplete.
+
+For basic blocks, I'm not maintaining functional correctness of the benchmarks. I'm replacing a basic block 
+with a floating point operator with a runtime function call. I'm collecting results to how much coverage
+this version has.
+
+For sub blocks, I tested the pass on a specific microbenchmark I wrote (`diamond`). However, this pattern
+does not appear elsewhere in MachSuite. 
+
+[mach]: https://breagen.github.io/MachSuite/
+
+## Performance
+First image illustrates performance at different abstraction levels for the `diamond` microbenchmark.
```

Comment:
Before getting into the results, maybe it would be useful to clarify what you were expecting to find? What would be a "good" result or a "bad" result? I think there are two dimensions here: correctness and performance. The "good" result for correctness may be obvious, but I'm not sure there is such a thing as a "good" outcome for performance.